### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by PatrickWatson887**

### CODE REVIEW

**Review by e-mccoy — APPROVED** [review]

File: Doras/src/app/band-competencies/band-competencies.component.spec.ts [hunk 1]
```diff
@@ -1,5 +1,19 @@
 import { async, ComponentFixture, TestBed } from '@angular/core/testing';
```
Comment (micalexis) [inline Doras/src/app/band-competencies/band-competencies.component.spec.ts:1]: fix the conflicts!!

**Review by micalexis — CHANGES_REQUESTED** [review]

File: Doras/src/app/app.component.html [hunk 2]
```diff
@@ -13,8 +13,8 @@
 
 <router-outlet></router-outlet>
 
-<div class="footer">
-<img src='assets/flux.jpg' width="100%" style="height: 50px;"/>
+<div class="footer" style="position: absolute; bottom: 0px; ">
+<img src='assets/flux.jpg' width="100%" style="height: 50px; bottom: 0px;"/>
```
Comment (micalexis) [inline Doras/src/app/app.component.html]: styling should be part of the css file for the component

File: Doras/src/app/carousel-tabs/carousel-tabs.component.ts [hunk 3]
```diff
@@ -1,36 +1,133 @@
...
+    this.getJobRoleCapId()
+ }
+
+ getCapabilities() {
+  this.subCapabilities = this.switchBoard.capability$.subscribe((c) => {
+    this.capabilities = c;
+  });
+ }
+ getJobRoleCapId() {
+  this.data.getJobRole().subscribe((c) => {
```
Comment (micalexis) [inline Doras/src/app/carousel-tabs/carousel-tabs.component.ts:41]: a, b, c??? name these variables according to what they actually are!

Comment (micalexis) [reply Doras/src/app/carousel-tabs/carousel-tabs.component.ts:41]: semicolons!!!!!!!! use eslint to fix these

File: Doras/src/app/carousel-tabs/carousel-tabs.component.ts [hunk 4]
```diff
@@ -1,36 +1,133 @@
...
+ getBand() {
+  this.data.getBand().subscribe(c => {
+    this.bands = c.reverse();
+    this.jobBandArray = this.splitInto3(this.bands)
+  });
+ }
+ getJobRoleBandId() {
+  this.data.getJobRole().subscribe((c) => {
+    this.otherBandArray = c
+    this.otherBandArray.sort((a, b) => (a.bandId > b.bandId) ? 1 : -1)
```
Comment (micalexis) [inline Doras/src/app/carousel-tabs/carousel-tabs.component.ts:73]: create a generic for method this is being repeated

File: Doras/src/app/carousel-tabs/carousel-tabs.component.ts [hunk 5]
```diff
@@ -1,36 +1,133 @@
...
 
-  onTabChange(tabChange: NgbTabChangeEvent) {
-    this.dataTransferService.capability = this.capabilities[tabChange.nextId];
+  onTabChange(event: MatTabChangeEvent) {
+    if (this.roles[event.index].length !== 3) {
+      for (let i = 0; i < this.bands.length; i++){
+        try {
+          if (this.bands[i].bandId != this.roles[event.index][i].bandId) {
+            this.roles[event.index].splice(i, 0, {
+              bandId: 9-i,
```
Comment (micalexis) [inline Doras/src/app/carousel-tabs/carousel-tabs.component.ts]: why the hard coded 9?! is this the number of bands? should we not be determining this from the number of bands retrieved from the db? and what happened to using  well named variables

File: Doras/src/app/carousel-tabs/carousel-tabs.component.ts [hunk 6]
```diff
@@ -1,36 +1,133 @@
...
   ngOnDestroy(): void {
     this.subCapabilities.unsubscribe();
  }
 
-  onTabChange(tabChange: NgbTabChangeEvent) {
-    this.dataTransferService.capability = this.capabilities[tabChange.nextId];
+  onTabChange(event: MatTabChangeEvent) {
+    if (this.roles[event.index].length !== 3) {
+      for (let i = 0; i < this.bands.length; i++){
+        try {
```
Comment (micalexis) [inline Doras/src/app/carousel-tabs/carousel-tabs.component.ts]: avoid the use of try catches to do conditional statements! its not a conditional statement!

File: Doras/src/app/carousel-tabs/carousel-tabs.component.ts [hunk 7]
```diff
@@ -1,36 +1,133 @@
...
+            })
+          }
+        } catch {
+          this.roles[event.index].push({
+            bandId: 9-i,
+            title: "No Role"
+          })
+        }
+      }
+      (this.roles[event.index].sort((a, b) => (a.bandId > b.bandId) ? 1 : -1)).reverse();
```
Comment (micalexis) [inline Doras/src/app/carousel-tabs/carousel-tabs.component.ts]: why? refer to comment above. leave comments explaining this

File: Doras/src/app/carousel-tabs/carousel-tabs.component.ts [hunk 8]
```diff
@@ -1,36 +1,133 @@
...
+      this.roles[event.index] = this.splitInto3(this.roles[event.index])
+
+    }
+    this.switchBoard.getJob(this.roles[event.index])
+    this.switchBoard.getBand(this.jobBandArray)
+    this.switchBoard.getOtherBand(this.otherBandArray)
+
+ }
+
+ splitInto3(array: any[]): any[] {
```
Comment (micalexis) [inline Doras/src/app/carousel-tabs/carousel-tabs.component.ts]: this should be called a generic name! and be able to pass in a value into step to dynamically group the results.

File: Doras/src/app/data-transfer.service.ts [hunk 9]
```diff
@@ -5,6 +5,15 @@ import { Capability } from './models/capability';
   providedIn: 'root'
 })
 export class DataTransferService {
-  capability: Capability | null;
+  private capability: Capability | null;
   constructor() { }
+
+  setCapability(capability: Capability) {
+    // console.log(capability)
```
Comment (micalexis) [inline Doras/src/app/data-transfer.service.ts]: remove

File: Doras/src/app/data.service.ts [hunk 10]
```diff
@@ -37,8 +37,8 @@ export class DataService {
     return this.http.get<JobFamily[]>('api/distinctJobFamilies');
```
Comment (micalexis) [inline Doras/src/app/data.service.ts:36]: do we have another call to bring back duplicate job families? this endpoint should just be /api/jobFamilies and you SQL should handle ensuring the results returned are unique job families

File: Doras/src/app/models/capability.ts [hunk 11]
```diff
@@ -1,6 +1,5 @@
 export class Capability {
-    id: number;
     name: string;
     leadId?: number;
```
Comment (micalexis) [inline Doras/src/app/models/capability.ts]: every capability must have a capability lead so why optional. is this column nullable in the database?

File: Doras/src/app/roles/roles.component.html [hunk 12]
```diff
@@ -6,54 +6,64 @@
...
-                <h1 *ngIf="job1; else notShowBlock">{{job1.title}}</h1>
-                <div *ngIf="job1;">
-                <p></p>
-                <img title="Responsibilities" src='assets/responsibilities.png' onmouseover="this.src='assets/responsibilities_hover.png';" onmouseout="this.src='assets/responsibilities.png';"class="topLeftIcon" (click)="openDialog('responsibility', 'job1')"/>
-                <img title="Training" src='assets/training.png' onmouseover="this.src='assets/training_hover.png';" onmouseout="this.src='assets/training.png';" class="topRightIcon" (click)="openDialog('training', 'training1')"/>
-                <p></p>
-                <img title="Job Specification" src='assets/jobSpec.png' onmouseover="this.src='assets/jobSpec_hover.png';" onmouseout="this.src='assets/jobSpec.png';" class="topLeftIcon" (click)="openDialog('spec', 'spec1')"/>
-                <img title="Other Jobs in the Same Band" src='assets/sjsb.png' onmouseover="this.src='assets/sjsb_hover.png';" onmouseout="this.src='assets/sjsb.png';" class="topRightIcon" />
-                </div>
+              <mat-card>
```
Comment (micalexis) [inline Doras/src/app/roles/roles.component.html:9]: this whole page is heavily hardcoded! each mat-card should be dynamically generated

File: Doras/src/app/roles/roles.component.ts [hunk 13]
```diff
@@ -1,72 +1,86 @@
...
+
+    this.subBand = this.switchBoard.band$.subscribe((c) => {
+      this.jobBandArray = c;
     });
+
+    this.subOtherBand = this.switchBoard.otherBand$.subscribe((c) => {
+      this.otherBands = c;
+    });
+
+    this.data.getCompetenciesBand(9).subscribe(c => {
```
Comment (micalexis) [inline Doras/src/app/roles/roles.component.ts:58]: what is going on here? 7,8,9? is there not a better way to do this? what are these values suppose to mean? 👎

File: Doras/src/app/specification/specification.component.html [hunk 14]
```diff
@@ -4,12 +4,12 @@ <h3 class="modal-title" id="modal-basic-title" mat-dialog-title>Specification</h
       <span aria-hidden="true">&times;</span>
     </button>
   </div>
-  <div class="modal-body" mat-dialog-content *ngFor="let job of data.data">
-    Title: {{ job.title }}
+  <div class="modal-body" mat-dialog-content>
+    Title: {{ data.data.title }}
```
Comment (micalexis) [inline Doras/src/app/specification/specification.component.html]: can we call is something else other than data?

**Review by micalexis — CHANGES_REQUESTED** [review]
a few changes as it is a big PR
serious reduction of hard coded values required
squash commits

**Pull request merged**